### DISCUSSION TITLE
API versioning in FASTAPI

### DISCUSSION THREAD

**Discussion by priya-gitTest** [fastapi/fastapi#8422] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
na
```

### Description

How are multiple versions of same api handled in Fastapi. what are the recomendations wrt project structure , if we can to acheive something like mentioned here : 
https://developers.facebook.com/docs/graph-api/guides/versioning/
https://stackoverflow.com/questions/64943693/what-are-the-best-practices-for-structuring-a-fastapi-project

### Operating System

macOS

### Operating System Details

na

### FastAPI Version

na

### Python Version

na

### Additional Context

https://stackoverflow.com/questions/64943693/what-are-the-best-practices-for-structuring-a-fastapi-project

**Comment by TheJumpyWizard** [comment]
Think this could be a good read for your question: https://medium.com/geoblinktech/fastapi-with-api-versioning-for-data-applications-2b178b0f843f @priya-gitTest

**Comment by extreme4all** [comment, marked as answer]
FastAPI's APIRouter class includes a prefix parameter that allows you to specify a prefix for all the routes defined in that router. This can be useful for organizing your API and for defining multiple versions of the same API.

Here's an example of how you might use the prefix parameter when defining a router in FastAPI:
```py
from fastapi import APIRouter, FastAPI

app = FastAPI()
router = APIRouter()

@app.get("/lol", tags=["users"])
async def lol():
    return {"detail": "lol"}

@router.get("/lol")
async def lol():
    return {"detail": "lol"}

app.include_router(
    router,
    prefix="/v1",
    tags=["users"],
)
```
![image](https://user-images.githubusercontent.com/40169115/205522067-005a5eba-f1ba-43ca-ace8-829dc6bfe7d2.png)

**Comment by extreme4all** [comment]
@priya-gitTest does this help you?

**Comment by priya-gitTest** [comment]
> https://medium.com/geoblinktech/fastapi-with-api-versioning-for-data-applications-2b178b0f843f

yes, I had ran into that blog to but forgot to add it here. Thanks for mentioning it.

**Comment by priya-gitTest** [comment]
> @priya-gitTest does this help you?

Yes, it does help to get it working but still unclear wrt project structure. Only reference I got is thi one but seems not very popular with Most FAST API crowd... https://christophergs.com/tutorials/ultimate-fastapi-tutorial-pt-8-project-structure-api-versioning/

**Comment by TheJumpyWizard** [comment]
@priya-gitTest it seems like extreme4all was suggesting what that christophergs.com tutorial suggested for api versioning. using version prefixes in the routes is a suitable initial method for versioning. If your question is more about the progression of api versioning in regards to deployments into an environment, and not just specifying the versioning as mentioned above, i would read up on canary deployments. api version releases can be deployed in numerous ways, and as an example, using something like Helm to specify the api version being deployed to a kubernetes cluster is one approach that could be good study material.
